packages/react/menu/src/Menu.tsx

```diff
@@ -438,16 +439,57 @@ const MenuContentImpl = React.forwardRef((props, forwardedRef) => {
     ...contentProps
   } = props;
   const context = useMenuContext(CONTENT_NAME);
-  const typeaheadProps = useMenuTypeahead();
   const { getItems } = useCollection();
   const [currentItemId, setCurrentItemId] = React.useState<string | null>(null);
   const contentRef = React.useRef<HTMLDivElement>(null);
   const composedRefs = useComposedRefs(forwardedRef, contentRef, context.onContentChange);
   const isPointerDownOutsideRef = React.useRef(false);
+  const timerRef = React.useRef(0);
+  const searchRef = React.useRef('');
+  const [isSearching, setIsSearching] = React.useState(false);
 
   const PortalWrapper = portalled ? Portal : React.Fragment;
   const ScrollLockWrapper = disableOutsideScroll ? RemoveScroll : React.Fragment;
 
+  // Reset `searchRef` 1 second after it was last updated
+  const handleSearchChange = React.useCallback((search: string) => {
+    searchRef.current = search;
+    setIsSearching(search.length > 0);
+    window.clearTimeout(timerRef.current);
+    timerRef.current = window.setTimeout(() => handleSearchChange(''), 1000);
+  }, []);
+
+  React.useEffect(() => {
+    return () => window.clearTimeout(timerRef.current);
+  }, []);
+
+  React.useEffect(() => {
+    const handleTypeaheadSearch = ((event: CustomEvent<{ key: string }>) => {
+      event.stopPropagation();
+      handleSearchChange(searchRef.current + event.detail.key);
```

Comment:
So this part has been bothering me a bit. I initially was going to suggest calling this after all the `const`s but then realised we can't do that because there is an implicit dependency on `searchRef.current` being updated first here so the `const nextMatch` can reference it.

The `handleSearchChange` seems to only exist so that it can call itself from the timeout and isn't being re-used anywhere so I was wondering what thoughts are on consolidating these two methods?

```jsx
const handleTypeaheadSearch = (key: string) => {
  const search = searchRef.current + key;
  const items = getItems().filter((item) => !item.disabled);
  const currentItem = document.activeElement;
  const currentMatch = items.find((item) => item.ref.current === currentItem)?.textValue;
  const values = items.map((item) => item.textValue);
  const nextMatch = getNextMatch(values, search, currentMatch);
  const newItem = items.find((item) => item.textValue === nextMatch)?.ref.current;

  (function updateSearch(search: string) {
    searchRef.current = search;
    setIsSearching(search.length > 0);
    window.clearTimeout(timerRef.current);
    timerRef.current = window.setTimeout(() => updateSearch(''), 1000);
  })(search);

  if (newItem) {
    /**
      * Imperative focus during keydown is risky so we prevent React's batching updates
      * to avoid potential bugs. See: https://github.com/facebook/react/issues/20332
      */
    setTimeout(() => (newItem as HTMLElement).focus());
  }
};
```

I ran this by @andy-hook already and we just wanted to check what your thoughts are @benoitgrelard really because we don't use IIFEs much.

Reply:
Went ahead and applied this – https://github.com/radix-ui/primitives/pull/662/commits/bb2f5f8f3026e811527eac47e3534e842c0b5be7 I personally prefer it